# <font color='red'><b>Bootstrap assignment</b> </font>

<b>There will be some functions that start with the word "grader" ex: grader_sampples(), grader_30().. etc, you should not change those function definition.

Every Grader function has to return True.</b>

<font color='blue'> <b>Importing packages</b> </font>

In [1581]:
import numpy as np # importing numpy for numerical computation
from sklearn.datasets import load_boston # here we are using sklearn's boston dataset
from sklearn.metrics import mean_squared_error # importing mean_squared_error metric
from sklearn.tree import DecisionTreeRegressor
import warnings
warnings.filterwarnings("ignore")

In [1582]:
boston = load_boston()
x=boston.data #independent variables
y=boston.target #target variable

In [1583]:
x.shape

(506, 13)

In [1584]:
x[:5]

array([[6.3200e-03, 1.8000e+01, 2.3100e+00, 0.0000e+00, 5.3800e-01,
        6.5750e+00, 6.5200e+01, 4.0900e+00, 1.0000e+00, 2.9600e+02,
        1.5300e+01, 3.9690e+02, 4.9800e+00],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, 0.0000e+00, 4.6900e-01,
        6.4210e+00, 7.8900e+01, 4.9671e+00, 2.0000e+00, 2.4200e+02,
        1.7800e+01, 3.9690e+02, 9.1400e+00],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, 0.0000e+00, 4.6900e-01,
        7.1850e+00, 6.1100e+01, 4.9671e+00, 2.0000e+00, 2.4200e+02,
        1.7800e+01, 3.9283e+02, 4.0300e+00],
       [3.2370e-02, 0.0000e+00, 2.1800e+00, 0.0000e+00, 4.5800e-01,
        6.9980e+00, 4.5800e+01, 6.0622e+00, 3.0000e+00, 2.2200e+02,
        1.8700e+01, 3.9463e+02, 2.9400e+00],
       [6.9050e-02, 0.0000e+00, 2.1800e+00, 0.0000e+00, 4.5800e-01,
        7.1470e+00, 5.4200e+01, 6.0622e+00, 3.0000e+00, 2.2200e+02,
        1.8700e+01, 3.9690e+02, 5.3300e+00]])

## <font color='red'><b>Task 1</b></font>

<font color='red'> <b>Step - 1</b></font>

*  <font color='blue'><b>Creating samples</b></font><br>
    <b> Randomly create 30 samples from the whole boston data points</b>
    *  Creating each sample: Consider any random 303(60% of 506) data points from whole data set and then replicate any 203 points from the sampled points
    
     For better understanding of this procedure lets check this examples, assume we have 10 data points [1,2,3,4,5,6,7,8,9,10], first we take 6 data points randomly , consider we have selected [4, 5, 7, 8, 9, 3] now we will replicate 4 points from [4, 5, 7, 8, 9, 3], consder they are [5, 8, 3,7] so our final sample will be [4, 5, 7, 8, 9, 3, 5, 8, 3,7]
* <font color='blue'><b> Create 30 samples </b></font>
    *  Note that as a part of the Bagging when you are taking the random samples <b>make sure each of the sample will have different set of columns</b><br>
Ex: Assume we have 10 columns[1 ,2 ,3 ,4 ,5 ,6 ,7 ,8 ,9 ,10] for the first sample we will select [3, 4, 5, 9, 1, 2] and for the second sample  [7, 9, 1, 4, 5, 6, 2] and so on...
Make sure each sample will have atleast 3 feautres/columns/attributes

* <font color='red'><b> Note - While selecting the random 60% datapoints from the whole data, make sure that the selected datapoints are all exclusive, repetition is not allowed. </b></font>

<font color='red'><b>Step - 2 </b></font>

<font color='blue'><b>Building High Variance Models on each of the sample and finding train MSE value</b></font>

*  Build a regression trees on each of 30 samples.
*  Computed the predicted values of each data point(506 data points) in your corpus.
*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{30}\sum_{k=1}^{30}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$
*  Now calculate the $MSE =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$

<font color='red'> <b>Step - 3 </b></font>

*  <font color='blue'><b>Calculating the OOB score </b></font>

*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{k}\sum_{\text{k= model which was buit on samples not included } x^{i}}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$.
*  Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

# <font color='red'><b>Task 2</b></font>

*  <font color='blue'><b>Computing CI of OOB Score and Train MSE</b></font>
  *   Repeat Task 1 for 35 times, and for each iteration store the Train MSE and OOB score </li>
<li> After this we will have 35 Train MSE values and 35 OOB scores </li>
<li> using these 35 values (assume like a sample) find the confidence intravels of MSE and OOB Score </li>
<li> you need to report CI of MSE and CI of OOB Score </li>
<li> Note: Refer the Central_Limit_theorem.ipynb to check how to find the confidence intravel</li>
</ol>

# <font color='red'><b>Task 3</b></font>

*  <font color='blue'><b>Given a single query point predict the price of house.</b></font>

Consider xq= [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60] 
Predict the house price for this point as mentioned in the step 2 of Task 1.

## <font color='red'><b>A few key points</b></font>

* Remember that the datapoints used for calculating MSE score contain some datapoints that were initially used while training the base learners (the 60% sampling). This makes these datapoints partially seen (i.e. the datapoints used for calculating the MSE score are a mixture of seen and unseen data).
Whereas, the datapoints used for calculating OOB score have only the unseen data. This makes these datapoints completely unseen and therefore appropriate for testing the model's performance on unseen data.

* Given the information above, if your logic is correct, the calculated MSE score should be less than the OOB score.

* The MSE score must lie between 0 and 10.
* The OOB score must lie between 10 and 35.

* The difference between the left nad right confidence-interval values must not be more than 10. Make sure this is true for both MSE and OOB confidence-interval values.

# <font color='blue'> <b>Task - 1</b></font>

<font color='blue'><b>Step - 1</b></font>

*  <font color='blue'> <b>Creating samples</b></font>

<font color='Orange'><b>Algorithm</b></font>

![alt text](https://i.imgur.com/OfcFrUP.jpg/)

*  <font color='blue'><b> Write code for generating samples</b></font>

In [1585]:
def generating_samples(input_data, target_data):

    '''In this function, we will write code for generating 30 samples '''
    # you can use random.choice to generate random indices without replacement
    # Please have a look at this link https://docs.scipy.org/doc/numpy-1.16.0/reference/generated/numpy.random.choice.html for more details
    # Please follow above pseudo code for generating samples 
    selecting_rows = np.random.choice(len(input_data), size=int(0.6*len(input_data)), replace=False)
    replacing_rows = np.random.choice(selecting_rows, size=int(0.4015*len(input_data)), replace=False)
    selected_columns = np.array(np.random.choice(a=input_data.shape[1], size=np.random.randint(3,input_data.shape[1]), replace=False))
    sample_data = input_data[selecting_rows[:, None], selected_columns]
    target_of_sample_data = target_data[selecting_rows]

    # Replacating Data
    replicated_sample_data = input_data[replacing_rows[:, None], selected_columns]
    target_of_replicated_sample_data = target_data[replacing_rows]

    # Concatinating data
    final_sample_data = np.vstack((sample_data, replicated_sample_data))
    final_target_data = np.vstack((target_of_sample_data.reshape(-1,1), target_of_replicated_sample_data.reshape(-1,1)))
    return final_sample_data, final_target_data, selecting_rows, selected_columns

    # return sampled_input_data , sampled_target_data,selected_rows,selected_columns
    # note please return as lists

<font color='cyan'> <b> Grader function - 1 </b> </fongt>

In [1586]:
def grader_samples(a,b,c,d):
    length = (len(a)==506  and len(b)==506)
    sampled = (len(a)-len(set([str(i) for i in a]))==203)
    rows_length = (len(c)==303)
    column_length= (len(d)>=3)
    assert(length and sampled and rows_length and column_length)
    return True
a,b,c,d = generating_samples(x, y)
grader_samples(a,b,c,d)

True

*  <font color='blue'> <b>Create 30 samples </b>

![alt text](https://i.imgur.com/p8eZaWL.jpg)

In [1587]:
# Use generating_samples function to create 30 samples 
# store these created samples in a list
list_input_data = []
list_output_data = []
list_selected_rows = []
list_selected_columns = []

for i in range(30):
    a, b, c, d = generating_samples(x, y)
    list_input_data.append(a)
    list_output_data.append(b)
    list_selected_rows.append(c)
    list_selected_columns.append(d)

<font color='cyan'> <b>Grader function - 2 </b></font>

In [1588]:
def grader_30(a):
    assert(len(a)==30 and len(a[0])==506)
    return True
grader_30(list_input_data)

True

<br>

<font color='red'><b>Step - 2 </b></font>

<font color='orange'><b>Flowchart for building tree</b></font>

![alt text](https://i.imgur.com/pcXfSmp.png)

*  <font color='blue'><b> Write code for building regression trees</b></font>

In [1589]:
def build_reg_tree(list_ip_data, list_op_data):
    """
    Function to build regressgion tree
    INPUT:
    list_ip_data = list of sample data created after generating samples
    list_op_data = list of target data created after generating samples

    OUTPUT:
    returns mutiple models after fiting data
    """
    models = []
    for i in range(30):
        dt = DecisionTreeRegressor()
        dt.fit(list_ip_data[i], list_op_data[i])
        models.append(dt)
    return models

In [1590]:
all_models = build_reg_tree(list_input_data, list_output_data)

<font color='orange'><b>Flowchart for calculating MSE </b></font>

![alt text](https://i.imgur.com/sPEE618.png)

After getting predicted_y for each data point, we can use sklearns mean_squared_error to calculate the MSE between predicted_y and actual_y.

*  <font color='blue'><b> Write code for calculating MSE</b></font>

In [1591]:
def calculate_mse(input_data, output_data, cols_selected, reg_trees):
    """
    Function to calculate MSE
    INPUT:
    input_data = independent variables of given data
    output_data = target variable of given data
    cols_selected = list of randomly selected column indices after random sampling
    reg_trees = multiple models of DecisionTree

    OUTPUT:
    returns mse value between actual_y and median of predicted_y's
    """
    array_of_Y = [] # list to store y_predicted

    for i in range(len(cols_selected)):
        # all data points for selected column indices
        datapoints = input_data[:, cols_selected[i]]
        # predict y 
        predicted_y = reg_trees[i].predict(datapoints)
        array_of_Y.append(predicted_y)

    array_of_Y = np.array(array_of_Y).T # Reshaping the array
    median = np.median(array_of_Y, axis=1) # median along axis=1

    return mean_squared_error(y, median) # MSE

In [1592]:
mse = calculate_mse(x, y, list_selected_columns, all_models)
print("Mean Squared Error : ", mse)

Mean Squared Error :  0.05849925889328065


<font color='blue'><b>Step - 3 </b></font>

<font color='orange'><b>Flowchart for calculating OOB score</b></font>

![alt text](https://i.imgur.com/95S5Mtm.png)

Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

*  <font color='blue'><b> Write code for calculating OOB score </b></font>

In [1593]:
def calculate_oob_score(input_data, output_data, rows_selected, cols_selected, reg_trees):
    """
    Function to calculate OOBScore
    INPUT:
    input_data = independent variables of given data
    output_data = target variable of given data
    rows_selected = list of randomly selected row indices after random sampling
    cols_selected = list of randomly selected column indices after random sampling
    reg_trees = multiple models of DecisionTree

    OUPUT:
    returns oob_score
    """
    final_median = [] # to store all median values

    for i in range(len(input_data)):
        y_scores = [] # to store all y_predicted values
        for idx in range(len(rows_selected)):
            # if a point(index) is not available in selected_rows
            if i not in rows_selected[idx]: 
                model = reg_trees[idx] # DT model
                # all row data points belonging to selected columns
                x_datapoints = input_data[i][cols_selected[idx]]
                x_datapoints = np.array(x_datapoints).reshape(1, -1)
                y_predict = model.predict(x_datapoints) # predict y
                y_scores.append(y_predict)
        y_scores_arr = np.array(y_scores)
        y_median = np.median(y_scores_arr) # median of y predicted
        final_median.append(y_median)

    # OOBScore
    temp_oob = 0
    for j in range(len(x)):
        temp_oob += ((output_data[j] - final_median[j]) ** 2)
    final_oob = temp_oob/len(output_data) # OOB score
    return final_oob

In [1594]:
oob_score = calculate_oob_score(x, y, list_selected_rows, list_selected_columns, all_models)
print("OOBScore : ", oob_score)

OOBScore :  15.092809427460043


# <font color='blue'><b>Task 2</b></font>

In [1595]:
def bagging(ip_data, op_data, n, m_samples):
    """
    Function to calculate MSE and OOB score multiple times
    INPUT:
    ip_data = independent variables of given data
    op_data = target variable of given data
    n = number of times bagging
    m_samples = number of times samples need to be generated

    OUTPUT:
    returns list of MSE values and list of OOB scores
    """
    mse_list = []
    oob_list = []
    # No. of times bagging 
    for times in range(n):
        list_selected_rows = []
        list_selected_columns = []
        list_input_data = []
        list_output_data = []
        # No. of times samples need to be generated
        for sample in range(m_samples):
            a,b,c,d = generating_samples(ip_data, op_data)
            list_input_data.append(a)
            list_output_data.append(b)
            list_selected_rows.append(c)
            list_selected_columns.append(d)

        # Build regression tree of 30 DT models
        all_models = build_reg_tree(list_input_data, list_output_data)
        # MSE value
        mse = calculate_mse(ip_data, op_data, list_selected_columns, all_models)
        # OOB score
        oob_score = calculate_oob_score(ip_data, op_data, list_selected_rows, list_selected_columns, all_models)
        mse_list.append(mse)
        oob_list.append(oob_score)
        
    return np.array(mse_list), np.array(oob_list)

In [1596]:
n = 35 # number of times task 1 is repeated
m_samples = 30 # number of times samples need to be generated
all_mse, all_oob = bagging(x, y, n, m_samples)

In [1597]:
def confidence_interval(list_of_val):
    """
    Function to calculate confidence interval
    INPUT:
    list_of_val = list of values generated after using some function

    OUTPUT:
    returns 95% confidence interval of input values
    """
    # mean of input values
    mean = np.round(np.mean(list_of_val), 3)
    # standard deviation of input values
    std_dev = np.round(np.std(list_of_val), 3)
    # lower limit confidence interval
    lower_limit = mean - 2 * (std_dev/np.sqrt(len(list_of_val))) 
    # upper limit of confidence interval
    upper_limit = mean + 2 * (std_dev/np.sqrt(len(list_of_val)))
    return [lower_limit, upper_limit]

In [1598]:
# confidence interval of MSE
ci_mse = confidence_interval(all_mse)
# confidence interval of OOB score
ci_oob = confidence_interval(all_oob)

In [1599]:
print("Confidence Interval of MSE :", ci_mse)
print("Confidence Interval of OOBScore :",ci_oob)

Confidence Interval of MSE : [0.07971391705222572, 0.16828608294777428]
Confidence Interval of OOBScore : [13.819552819521036, 14.884447180478965]


# <font color='blue'><b>Task 3</b></font>

<font color='orange'><b>Flowchart for Task 3</b></font>

<b>Hint: </b> We created 30 models by using 30 samples in TASK-1. Here, we need send query point "xq"  to 30 models and perform the regression on the output generated by 30 models.

![alt text](https://i.imgur.com/Y5cNhQk.png)

*  <font color='blue'><b> Write code for TASK 3 </b></font>

In [1600]:
xq = [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60]

predicted_Y = [] # list to store y_predicted

for i in range(len(list_selected_columns)):
    # all data points for selected column indices
    datapoints = [xq[col] for col in list_selected_columns[i]]
    # predict y 
    predicted_y = all_models[i].predict(np.array((datapoints)).reshape(1,-1))
    predicted_Y.append(predicted_y)

predicted_Y = np.array(predicted_Y)
median = np.median(predicted_Y) # median of predicted values

In [1601]:
print("Predicted House Price :", median)

Predicted House Price : 18.5


<font color='red'><b>Write observations for task 1, task 2, task 3 indetail</b></font>

**Task 1** :
- We generated 30 samples from input data using row sampling and column sampling.
- MSE values are in the range of 0-0.5 every runtime.
- OOB Scores are in the range of 10-20 every runtime.
- MSE value varied a lot between `[0.002 to 0.211]` every runtime(22 tests):
    - 3/22 cases value was lower than `0.01`.
    - 8/22 cases value was higher than `0.1`.
    - 11/22 cases value was between `[0.02 to 0.099]`
- OOB Score varied between `[10 to 20]` every runtime(22 test):
     - On average the value was around `15`.
     - 5/22 cases value was lower than `13`.
     - 2/22 cases value was higher than `17`.
     - 15/22 cases value was between `[13 to 17]`
- From these observations:
    - MSE values are a little unreliable as they tend to vary a lot for the same type of dataset.
    - OOB scores are fairly reliable as they don't tend to vary as much.

**Task 2** :
- Here our MSE and OOB score range were pretty consistent. The difference between lower and upper limit for MSE was on average **`0.0575`** and for OOB score the difference was on average **`1.15`** in 95% of the cases.
- Only in 1/22 cases MSE value and OOB score was much higher than average **`0.0575`** and **`1.15`** respectively.
- 95% of the MSE and OOB score intervals calculated would contain the population mean.

**Task 3** :
- In task 3 we performed regression on the output generated by 30 DecisionTree models from task 1 which gave us the predicted value as `18.5` most of the time.
- Predicted value varied between `18.5` to `19.5`.
- Predicted value was higher than `18.5` when:
    - OOB score was higher than `17`.
    - In some cases OOB score was lesser than `13`.
    - In some cases where MSE was very low(`lesser than 0.02`) and OOB score was either lower(`lesser than 14`) or higher(`greater than 17`) than average OOB score(`around 15`)
- In cases where MSE was higher(`greater than 0.1`) and OOB score was lower(`lesser than 13`) the predicted value was either `18.5` or very close to it.
- After calculating MSE and OOB score only once to generalize the result may mislead our interpretation of the model performance. So finding confidence interval for MSE and OOB score seems to be a good choice as it will give us an average value for both which we can then use to interpret the model, whether it is performing well or not.
